### PR TITLE
Fixes #25832 - Show custom syspurpose values in web UI

### PR DESCRIPTION
**Pull request by jturel**
previous behavior: the dropdowns for syspurpose Addons, Usage and Roles wouldn't show any custom value (not derived from the manifest) which was set on the host. (Though it would show when the field is not in edit mode)

new behavior: the custom value will show in the dropdown/multiselect when set.
bonus: the org would be fetched every time the edit mode was invoked for all syspurpose fields - now the org is cached to make things much snappier

to test:

set a custom syspurpose attribute value on a content host:

```
facet = ::Host.find(1).subscription_facet
facet.purpose_role = 'customrole'
facet.purpose_usage = 'customusage'
facet.purpose_addons = ['customaddon']
facet.save!
```

look at the Content Host details view for the host and observe that these custom values aren't in the dropdowns when setting a new value.

check out this PR and observe that the custom values *do* show in the dropdowns or multiselect. also note that the organizations API is called only once with the new code

### CODE REVIEW

**Comment by jturel** [comment]
There are a few methods in this controller which repeat a similar block of code:

```
Organization.get({id: CurrentOrganization}, function (organization) {
    deferred.resolve(organization.system_purposes.roles);
});
```
Each time the relevant dropdowns are opened the org is fetched via API and I was trying to figure out a way to memoize the org but am struggling with the promise aspect. I tried something like this:
```
$scope.organization = null;

$scope.getOrganization = function() {
    if (!$scope.organization) {
      $scope.organization = Organization.get({id: CurrentOrganization}, function (organization) {
          return organization;
       });
     }
     return $scope.organization;
};
```

However, it doesn't seem to work when I use it:

```
$scope.getOrganization().promise.then(function(org) {
  // org is undefined here
});
```

@johnpmitsch I'm probably neglecting something obvious here - does anything jump out at you?

**Comment by jturel** [comment]
Didn't push it up yet but I was able to get it working like this:

```
        $scope.purposeRoles = function () {                                                                                                                                  
            return $scope.getOrganization().$promise.then(function(org) {                                                                                                    
                var roles = org.system_purposes.roles;                                                                                                                       
                var role = $scope.host.subscription_facet_attributes.purpose_role;                                                                                           
                if (role && !_.includes(roles, role)) {                                                                                                                      
                    roles.push(role);                                                                                                                                        
                }                                                                                                                                                            
                return roles;                                                                                                                                                
            });                                                                                                                                                              
        };
```

**Comment by johnpmitsch** [comment]
It looks like a step in the right direction, I have to look further tomorrow and maybe get some more context for the issue, but my first thought is I don't see a huge point for having a `$scope.getOrganization` function. My understanding of things is `Organization` is an angular factory [we create](https://github.com/Katello/katello/blob/master/engines/bastion_katello/app/assets/javascripts/bastion_katello/organizations/organization.factory.js) as a [Bastion Resource](https://github.com/Katello/bastion/blob/master/app/assets/javascripts/bastion/utils/bastion-resource.factory.js) which we ultimately returns a [$resource](https://docs.angularjs.org/api/ngResource/service/$resource) that provides the default method of `get`. This should return a promise, so you should be able to use this on its own. I think what you are trying to do is attach the organiation to the `$scope`, so we don't keep calling it. You could do something like this:

```javascript
If (!$scope.organization) {
  Organization.get({id: CurrentOrganization})
                      .then(function(org) {
                        $scope.organization = org
                      }).catch(function(err) {
                         // handle error (not entirely sure about this part yet or if its needed, the angular docs don't seem to do this)
                      })
}
```

And then proceed to just use `$scope.organization` instead of more promises. Hopefully this makes sense, this is just my first thought and the way you have it may be fine too. I can look more with you later

**Comment by jturel** [comment]
Yep, I'm trying to avoid multiple requests to fetch the org which implies attaching it to scope as far as I can tell.

With the approach you suggest is it guaranteed that $scope.organization would not be undefined by the time any `$scope.organization.$promise.then....` is called? The getOrganization approach guarantees that it won't be undefined

**Comment by johnpmitsch** [comment]
@jturel Would something like this work? I'm just not sure I understand the need to make `getOrganization` a promise that returns another promise, but I don't think I fully understand the issue. Let me know what you think
```javascript
// put on top of module
if (!$scope.organization) {
  Organization.get({id: CurrentOrganization})
              .then(function(org) { 
                $scope.organization = org
               })
}

// follow this kind of pattern when you need organization
$scope.serviceLevels = function () {
    if ($scope.organization) {
        return org.service_levels;
    });
};
```

**Comment by jturel** [comment]
@johnpmitsch the idea of using the getOrganization is so that org is memoized - we only have to fetch it once. It also means we don't have to check if it's defined - we can just depend on getOrganization to do the right thing. Here's a ruby example:

```
def my_instance_method
  @instance_variable ||= SomeService.get_response
end
```

By calling `my_instance_method` I never need to worry about getting nil, and I don't care about `@instance_variable` storing the value. This was my thinking with this approach.

I see what you're getting at with your approach too, but $scope.organization needs to return a promise so that the methods using it can have a guarantee that they're looking at an Organization otherwise the 4 methods calling it could break. Pushing up a new commit, take a look!

**Comment by jturel** [comment]
btw this PR is easier to visualize with 'Hide whitespace changes' selected!

**Comment by johnpmitsch** [comment]
@jturel Its looking good, I'm just hung up on the need of promises inside the functions. Once you have the organization, it seems like you shouldn't need promises anymore. (I'm referring to the `$promise.then` parts). 

Here is something I tried and it seemed to work well. You may need to modify it as I didn't test every situation and linting will most likely complain, but the idea is just check for `$scope.organization` before grabbing it's attributes

```diff
diff --git a/engines/bastion_katello/app/assets/javascripts/bastion_katello/content-hosts/details/content-host-details.controller.js b/engines/bastion_katello/app/assets/javascripts/bastion_katello/co
index 43a535e..24c54fa 100644
--- a/engines/bastion_katello/app/assets/javascripts/bastion_katello/content-hosts/details/content-host-details.controller.js
+++ b/engines/bastion_katello/app/assets/javascripts/bastion_katello/content-hosts/details/content-host-details.controller.js
@@ -154,14 +154,13 @@ angular.module('Bastion.content-hosts').controller('ContentHostDetailsController
         };
 
         $scope.purposeUsages = function () {
-            return $scope.organization.$promise.then(function(org) {
-                var usages = org.system_purposes.usage;
-                var usage = $scope.host.subscription_facet_attributes.purpose_usage;
-                if (usage && !_.includes(usages, usage)) {
-                    usages.push(usage);
-                }
-                return usages;
-            });
+            if (!$scope.organization) return []
+            var usages = $scope.organization.system_purposes.usage;
+            var usage = $scope.host.subscription_facet_attributes.purpose_usage;
+            if (usage && !_.includes(usages, usage)) {
+                usages.push(usage);
+            }
+            return usages;
         };
 
         $scope.purposeAddons = function () {
```

Also you may find lodash's [memoize](https://lodash.com/docs/4.17.11#memoize) function helpful, but honestly, I like how the organization fetching part is currently set up.

Let me know any thoughts or questions, I think there are a few ways to go about this.

**Comment by jturel** [comment]
I don't think I've explained the need very well so I am going to try again.  Sorry if the following sounds redundant to some of my other comments but I can't seem to find other language.

The main idea is that we are dealing with asynchronous behavior here because the organization is loaded via HTTP request. 

Looking at one method for example, `purposeUsages`, it's feasible that this method could be called (by expanding the dropdown on the UI) *before* the organization has been loaded (maybe the server is busy and running slow). This method only has meaning when we have an organization. We can more or less guarantee that it has an org by using the callback `$scope.organization.$promise.then` so that *when* we have the org we can return the expected list of values. I think this is a textbook example of using a promise unless I'm missing something and I could very well be 😆 

Your example is not really a substitute because it assumes the organization API has returned and reached the client. Though, the empty set return could be feasible for a 'catch' to prevent the relevant dropdown  from blowing up!

File: engines/bastion_katello/app/assets/javascripts/bastion_katello/content-hosts/details/content-host-details.controller.js [hunk 1]
```diff
@@ -24,6 +24,10 @@ angular.module('Bastion.content-hosts').controller('ContentHostDetailsController
         $scope.getHostStatusIcon = ContentHostsHelper.getHostStatusIcon;
         $scope.getHostPurposeStatusIcon = ContentHostsHelper.getHostPurposeStatusIcon;
 
+        if (!$scope.organization) {
```
Comment (jturel) [inline engines/bastion_katello/app/assets/javascripts/bastion_katello/content-hosts/details/content-host-details.controller.js]: I'm going to test it out - but can you think of any reason to not remove the if statement around this assignment? I don't think it's needed (it hung around from a previous commit)

File: engines/bastion_katello/app/assets/javascripts/bastion_katello/content-hosts/details/content-host-details.controller.js [hunk 2]
```diff
@@ -24,6 +24,8 @@ angular.module('Bastion.content-hosts').controller('ContentHostDetailsController
         $scope.getHostStatusIcon = ContentHostsHelper.getHostStatusIcon;
         $scope.getHostPurposeStatusIcon = ContentHostsHelper.getHostPurposeStatusIcon;
 
+        $scope.organization = Organization.get({id: CurrentOrganization});
+
```
Comment (jturel) [inline engines/bastion_katello/app/assets/javascripts/bastion_katello/content-hosts/details/content-host-details.controller.js]: @johnpmitsch turns out the surrounding `if` wasn't needed.

the execution of the code is like this:

- load content host details page
- controller JS is executed and $scope.organization becomes an angular resource - whether or not the server has returned is another story
- open dropdown for any of the syspurpose fields, loading is deferred until we have an org (and ideally we will by then!)

I verified that we are still making only 1 request to /organizations/:id across the different dropdowns

**Comment by jturel** [comment]
latest push broke a test ... looking into it

File: engines/bastion_katello/app/assets/javascripts/bastion_katello/content-hosts/details/content-host-details.controller.js (hunk 2 above)
Comment (johnpmitsch) [reply engines/bastion_katello/app/assets/javascripts/bastion_katello/content-hosts/details/content-host-details.controller.js]: Thanks for testing it out! Maybe we are starting to actually use `$resource` correctly? :thinking:

Comment (jturel) [reply engines/bastion_katello/app/assets/javascripts/bastion_katello/content-hosts/details/content-host-details.controller.js]: better late than never!

**Comment by jturel** [comment]
@johnpmitsch tests added!

**Pull request merged**